BUG: Breaking change to offsets.pyx - MonthOffset

Pawel-Kranzberg opened the pull request:
Add ability to use class `MonthBegin` as a period.
It would address #38914 and #38859.

- [ ] closes #38914 #38859
- [x] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Pawel-Kranzberg committed:
Update offsets.pyx

Fixes to year, quarter and month offsets.

Pawel-Kranzberg committed:
Merge remote-tracking branch 'upstream/master'

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Minor formatting changes.

Pawel-Kranzberg committed:
Update offsets.pyx

Minor formatting changes.

Pawel-Kranzberg committed:
Add ability to provide months as short names

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Temporary unfixing of the default months for `QuarterBegin` and `BQuarterBegin`.

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
[BUG] Breaking change to offsets.pyx - MonthOffset

Add ability to use `MonthBegin`, `BMonthBegin` and `BMonthEnd` as periods, e.g., by using their offset aliases as `freq` in `date_range`.
It would address #38914.

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

jreback reviewed (changes requested):
can you add a test for this that fails on current master.

you also need to adjust tests for this change.

jbrockmendel commented:
im wary of this.  inevitably someone is going to be confused as to why `Period(foo, freq= BusinessMonthEnd)` isnt different from `Period(foo, freq= BusinessMonthBegin)`

If we were doing this from scratch, I would keep DateOffsets totally unrelated to PeriodDtypes.

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update test_asfreq.py

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg commented:
> im wary of this. inevitably someone is going to be confused as to why `Period(foo, freq= BusinessMonthEnd)` isnt different from `Period(foo, freq= BusinessMonthBegin)`
> 
> If we were doing this from scratch, I would keep DateOffsets totally unrelated to PeriodDtypes.

I've removed changes related to `BusinessMonthBegin` and `BusinessMonthEnd` (still, it might make sense to create a proper business month period).

Pawel-Kranzberg committed:
Update offsets.pyx

Rollback of _lite_rule_alias change.

Pawel-Kranzberg committed:
Update dtypes.pyx

Add `"MS"` as a monthly code to `_period_code_map`

Pawel-Kranzberg committed:
Update test_asfreq.py

Pawel-Kranzberg committed:
Update test_asfreq.py

Pawel-Kranzberg committed:
Update test_asfreq.py

Pawel-Kranzberg committed:
Update test_asfreq.py

Pawel-Kranzberg commented:
@jbrockmendel - A question: would you prefer `Period(foo, freq = "M") != Period(foo, freq = "MS")` 
or `Period(foo, freq = "M") == Period(foo, freq = "MS")`? 
Assuming that in both cases
 `Period(foo, freq = "M").to_timestamp(how="start") == Period(foo, freq = "MS").to_timestamp(how="start")` 
 and  `Period(foo, freq = "M").to_timestamp(how="end") == Period(foo, freq = "MS").to_timestamp(how="end")`.
The former is easy, and that's how it is in the current version of this PR. The latter would require more analysis on my part, as my first approach at it through the `_lite_rule_alias` `dict` clashes with `DatetimeIndex`.

Pawel-Kranzberg committed:
Update test_period.py

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Update offsets.pyx

jreback commented on pandas/tests/scalar/period/test_asfreq.py:
```diff
@@ -791,14 +790,11 @@ def test_asfreq_combined(self):
 
     def test_asfreq_MS(self):
         initial = Period("2013")
```
can you parameterize this test instead with these different cases

Pawel-Kranzberg replied:
I did, please have a look at it.

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -2208,15 +2209,17 @@ cdef class BusinessMonthEnd(MonthOffset):
     Examples
     --------
-    >>> from pandas.tseries.offset import BMonthEnd
+    >>> from pandas.tseries.offsets import BusinessMonthEnd
```
is this just a typo in the example?

Pawel-Kranzberg replied:
Yes.

jreback commented on pandas/_libs/tslibs/dtypes.pyx:
```diff
@@ -78,6 +78,7 @@ _period_code_map = {
     "Q-NOV": 2011,    # Quarterly - November year end
 
     "M": 3000,        # Monthly
+    "MS": 3000,       # Monthly
```
why do we think adding this here is a good idea? this makes the map not invertible any more.

Pawel-Kranzberg replied:
I'll rework it, but probably in the second half of the week.

Pawel-Kranzberg replied:
fixed

jbrockmendel commented:
>  A question: would you prefer Period(foo, freq = "M") != Period(foo, freq = "MS")
or Period(foo, freq = "M") == Period(foo, freq = "MS")?

I think I would prefer that `Period(foo, freq="MS")` continue to raise.

The connection between DateOffset subclasses and valid Period freqs is basically a coincidence.  There is not a 1-1 mapping between them, and trying to shoehorn one in is going to make this less clear to people.

Pawel-Kranzberg committed:
Update dtypes.pyx

Add improved "MS" (month start) entry to _period_code_map

Pawel-Kranzberg committed:
Update test_asfreq.py

Parametrize and rename `test_M_vs_MS`, add `MS` tests to `test_asfreq_near_zero`, `test_conv_weekly` and `test_conv_monthly`.

Pawel-Kranzberg committed:
fix formatting

Pawel-Kranzberg committed:
Update test_asfreq.py

Fix test_asfreq_M_vs_MS

jbrockmendel commented on pandas/_libs/tslibs/dtypes.pyx:
```diff
@@ -77,7 +77,8 @@ _period_code_map = {
 
-    "M": 3000,        # Monthly
+    "M": 3000,        # Monthly - month end
+    "MS": 3001,       # Monthly - month start
```
the integers here are semantically meaningful, 3001 doesnt make sense in this context

Pawel-Kranzberg replied:
OK, I'll try to obtain a better number. It would help if I knew the meaningful relation between "M" and 3000. If the last zero symbolises the end (of month), then by some convoluted logic maybe one could represent the beginning, though?

jbrockmendel replied:
If the idea is for "MS" to map to the same type of Period as "M", then just use 3000.  If the idea is to have a Period that specifically corresponds to "Start of the Month" rather than "Month", then that really isn't what Period represents.

Pawel-Kranzberg replied:
I don't think using 3000 for both "M" and "MS" would be correct - please have a look at this screenshot:
![image](https://user-images.githubusercontent.com/7954786/112280268-1f623180-8c85-11eb-84eb-042fb0566d3a.png)

jbrockmendel replied:
there are aliases already in _period_code_map, but they are added on L104 and L108 after _reverse_period_code_map is defined.

Pawel-Kranzberg replied:
Done.

Pawel-Kranzberg committed:
Fixed "MS" code mapping in _period_code_map

Pawel-Kranzberg committed:
Merge remote-tracking branch 'upstream/master'

Pawel-Kranzberg committed:
Merge branch 'master' into patch-1

Pawel-Kranzberg committed:
Add PeriodDtypeCode to MonthBegin

Pawel-Kranzberg committed:
Update offsets.pyx

Pawel-Kranzberg committed:
Fix MonthBegin PeriodDtypeCode

Pawel-Kranzberg committed:
Bug hunt

Pawel-Kranzberg committed:
Remove double BusinessMonthBegin

Pawel-Kranzberg committed:
Merge pull request #1 from pandas-dev/master

Sync with pandas-dev/pandas master

Pawel-Kranzberg committed:
Merge pull request #4 from pandas-dev/master

sync with master

Pawel-Kranzberg committed:
Coverage bug hunt

Alter _period_code_map

Pawel-Kranzberg committed:
Revert "Coverage bug hunt"

This reverts commit 0628d22d7fb04b3632a326a7384d11a1ac91fa57.

Pawel-Kranzberg committed:
Bug hunt

Pawel-Kranzberg committed:
Revert "Bug hunt"

This reverts commit 56be0c9efe0679b980e54051d63892e23b6b886d.

Pawel-Kranzberg committed:
retrigger checks

Pawel-Kranzberg committed:
Update v1.3.0.rst

Pawel-Kranzberg committed:
Merge branch 'master' into patch-1

Pawel-Kranzberg committed:
Update v1.3.0.rst

Pawel-Kranzberg committed:
retrigger checks

Pawel-Kranzberg committed:
Merge branch 'patch-1' of https://github.com/Pawel-Kranzberg/pandas into patch-1

Pawel-Kranzberg commented:
> can you add a test for this that fails on current master.
> 
> you also need to adjust tests for this change.

done

jbrockmendel commented:
I find this really sketchy.  MonthBegin doesn't correspond to a Period.  I just dont understand the motivation here.

simonjayhawkins commented:
@Pawel-Kranzberg can you address https://github.com/pandas-dev/pandas/pull/39887#issuecomment-828108580

Pawel-Kranzberg commented:
> @Pawel-Kranzberg can you address [#39887 (comment)](https://github.com/pandas-dev/pandas/pull/39887#issuecomment-828108580)

I plan to do it next week. In a nutshell, there are existing issues that reference the need for `MonthBegin` as `Period` and I also faced it during a project at work. Ended up coding a workaround that made code maintenance more challenging.

mroeschke commented:
Thanks for the PR, but it appears that it has gotten stale and there's some uncertainty that around this proposed change.

Closing, but glad to have you help on other open issues.
